android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/MainActivity.java

```diff
@@ -548,52 +550,67 @@ public void onKeyboardDismissed() {
    */
   private void downloadKMP() {
     Intent downloadIntent;
-    String url = data.getQueryParameter(KMKeyboardDownloaderActivity.KMKey_URL);
-    if (url == null) {
-      url = data.toString();
-    }
-    if (url != null) {
-      // URL contains KMP to download in background.
-      boolean isCustom = FileUtils.isCustomKeyboard(url);
-
-      String filename = data.getQueryParameter("filename");
-      if (filename == null) {
-        int index = url.lastIndexOf("/") + 1;
-        if (index >= 0 && index <= url.length()) {
-          filename = url.substring(index);
-        }
+    try {
+      String url = data.getQueryParameter(KMKeyboardDownloaderActivity.KMKey_URL);
+      if (url == null) {
+        url = data.toString();
       }
+      if (url != null) {
+        // URL contains KMP to download in background.
+        boolean isCustom = FileUtils.isCustomKeyboard(url);
+
+        String filename = data.getQueryParameter(KMKeyboardDownloaderActivity.KMKey_Filename);
+        if (filename == null) {
+          int index = url.lastIndexOf("/") + 1;
+          if (index >= 0 && index <= url.length()) {
+            filename = url.substring(index);
+          }
+        }
 
-      // Only handle ad-hoc kmp packages or from keyman.com
-      if (FileUtils.hasKeymanPackageExtension(url) || data.getScheme().toLowerCase().equals("keyman")) {
-        try {
-          // Download the KMP to app cache
-          downloadIntent = new Intent(MainActivity.this, DownloadIntentService.class);
-          downloadIntent.putExtra("url", url);
-          downloadIntent.putExtra("filename", filename);
-          downloadIntent.putExtra("destination", MainActivity.this.getCacheDir().toString());
-          downloadIntent.putExtra("receiver", resultReceiver);
-
-          progressDialog = new ProgressDialog(MainActivity.this);
-          String ellipsisStr = "\u2026";
-          progressDialog.setMessage(String.format("%s\n%s%s",
-            getString(R.string.downloading_keyboard_package), filename, ellipsisStr));
-          progressDialog.setCancelable(false);
-          progressDialog.show();
-
-          startService(downloadIntent);
-        } catch (Exception e) {
-          if (progressDialog != null && progressDialog.isShowing()) {
-            progressDialog.dismiss();
+        String languageID = data.getQueryParameter(KMKeyboardDownloaderActivity.KMKey_Language);
+        String languageStr = (languageID != null) ? String.format("&language=%s", languageID) : "";
+        url = String.format("%s%s", url, languageStr);
+
+        String platform = data.getQueryParameter("platform");
+        String platformStr = (platform != null) ? String.format("&platform=%s", platform) : "";
+        url = String.format("%s%s", url, platformStr);
```

Comment:
So we reconstruct the URL, rather than use what was provided to us?  Is this so we can filter out any unwanted URL components or something?

I see that we do use the `languageID` variable later, which is what I assume preserves it for installation.

Reply:
From the test link I was using:
> keyman://?filename=sil_cameroon_qwerty.kmp&url= **https://keyman.com/keyboard/download?id=sil_cameroon_qwerty** &language=bbj-Latn&platform=android

line 554 returned the bolded portion for `url` (ignore the extra spacing I added for formatting.

Reply:
But we're assigning values to `url` here and using them later - like with line 583.  Why not just use the initial value of `url` (line 558) instead?  That's what I'm asking about.

Reply:
ah, now I remember why I had to append to url. The download endpoint will fail if I don't include `&platform=android`.

> Usage: download.php?id=<keyboard_id>&platform=<platform>[&mode=<bundle|standalone>][&cid=xxxx]

Reply:
If you could add a comment to help make that explicit when doing future maintenance, that'd be great.  Thanks for the explanation.